### PR TITLE
Stats v4: fix the time range for "This Week" tab considering daily saving time change

### LINKED ISSUE #2051: Store stats: "This Week" stats broken when device set to UK English
**Describe the bug**

With my device language set to English and device region set to United Kingdom, the "This Week" tab loads the wrong week. The dates for This Week appear as "30 Mar-30 Mar" when that's next week, not this week.

When I change my device region to United States, this week loads as expected (Mar 22-Mar 28).

**To Reproduce**

1. Go to the Settings app > General > Language & Region > Region.
2. Select United Kingdom and confirm the change.
3. Relaunch the Woo app.
4. Make sure improved stats are enabled (under Settings > Experimental features).
5. Go to the My store tab and select This Week.

**Screenshots**

US English (expected)|UK English (broken)
-|-
 | 

**Mobile Environment**
Please include:
- Device: iPhone XS
- iOS version: 13.3.1
- WooCommerce iOS version: 3.9.0.1

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #2051

## Cause of the issue

Currently, we calculate the time range for a stats v4 tab by:
- Calculate the latest date of the time range by the current date
- Calculate the earliest date of the time range by the latest date

In the case of "This Week" tab, the issue with daily saving time change goes like:
- Current date: March 29, 2020 (Sunday) / time zone: Rome, Italy (a week starts on Monday)
- Calculate the latest date of time range by the current date: we call `currentDate.endOfWeek(timezone:)`, which first calculates the `currentDate.startOfWeek(timezone:)` (March 23, Monday), then adds **`604799` seconds (7 days - 1 second)** to this start of week date.
  - This would normally work, because a week is normally 7 days of 24 hours. However, on a day with daily saving time change, a week is no longer 604800 seconds and could be +- 1 hour (or who knows how countries change the rules) --> this results in the latest date to become March 30, 2020, 0:59:59am
- Calculate the earliest date of the time range by the latest date --> because the latest date was incorrectly calculated as March 30, 2020 (Monday), the earliest date is also March 30, 2020 because the week starts on Monday.

## Changes

- When calculating the end of week for a date, instead of adding a fixed time interval (**`604799` seconds (7 days - 1 second)**) to the start of the week, now it adds one week by using `calendar.date(byAdding: .weekOfYear, value: 1, to: weekStartDate)` to find the date of the next week start date, then minus 1 second.
- Added 2 test cases for `startOfWeek` and `endOfWeek` for a Sunday with and without daily saving time change

## Testing

Prerequisites:
1. set the device date time to anytime on March 29, 2020 in Rome, Italy (or any other city that has daily saving time change on the same day) under Settings > "Date & Time".
2. the site has WC 4.0+ or WC Admin plugin activated

Notes: try relaunching the app once if the time range doesn't look correct, since the site time zone is refreshed on switching store or app launch.

### Case 1: Site with time zone in UTC offset

- In wp-admin, set the site time zone to **a UTC offset** (try a non-integer one) under Settings
- Launch the app with stats v4 enabled (Settings > Experimental Features > Stats) --> the time range on each tab (today, this week, this month, this year) should look correct

### Case 2: Site with time zone in a region with daily saving time change on the same day

- In wp-admin, set the site time zone to a **Rome** under Settings
- Launch the app with stats v4 enabled (Settings > Experimental Features > Stats) --> the time range on each tab (today, this week, this month, this year) should look correct

### Case 3: Site with time zone in a region without daily saving time change on this day

- In wp-admin, set the site time zone to **New York** under Settings
- Launch the app with stats v4 enabled (Settings > Experimental Features > Stats) --> the time range on each tab (today, this week, this month, this year) should look correct

## Example screenshots

before | after
-- | --
IMG_7686 | IMG_7685

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by jaclync** [comment]
> I think I see the wrong info. From your description and from your screenshots, I've to see the Mar 30 as starting date, and Apr 5 as end date, but in all the 3 cases, I see the range Mar 29-Apr4.
I also don't understand why in your screenshots before and after shows the same date (the only thing that seems changed is the time of the device).

thanks for testing @pmusolino !

did you test it on a simulator? I think Xcode has the default app region that has the week starting on Sunday, while my test device with Rome/Italy time zone has the week starting on Monday. Maybe if you change the "Application Region" to Rome/Italy in Xcode build options it might change the calendar?

Screen Shot 2020-04-10 at 12 27 27 PM

**Comment by jaclync** [comment]
@pmusolino I realized my example screenshots weren't correct, since I set the time back to March 29 and confused the screenshots from the future on the device 🤦🏻‍♀️just updated them in the PR description!

> the week showed will be 23 Mar-29 Mar which is pretty strange.
On the web, I selected as timezone: -5:30. I obtain the same result with the timezone configured to 5:30.
Using your timezone, I see the same data as you. I'm thinking, are you sure you put the Italian time zone during your tests?

yes, my device was set to March 29 in Rome/Italy for the screenshots. `23 Mar-29 Mar` for "This week" tab looks correctly, if the device time is on March 29. since the first weekday starts on Monday in Rome/Italy calendar, it ends on Sunday (March 29).

> Another thing that I tried this time, is to run DateStartAndEndTests. Seems that in my case I obtain 6 failing tests:

great catch! after changing the language region, I was able to reproduce the test failures. It looks like it's from different first weekday **based on the app language region not the time zone** (e.g. Italy starts on Monday vs. Taipei starts on Sunday - can be verified in the Apple Calendar app). I just [added a commit](https://github.com/woocommerce/woocommerce-ios/pull/2098/commits/f2c04673f7a9e7861f4bf967ba1db92d2088ea97) that passed the locale to the start/end of week date calculation and updated unit tests with locales that have different first weekday for another look!

**Comment by oguzkocer** [comment]
Hi there! `4.1` is being frozen, so this PR is being bumped to `4.2`. If these changes _have to_ be shipped with `4.1`, please target `release/4.1` and cc me in the PR so I can cut a new beta release. Also if you can keep the release dates in mind and bump the PRs & issues you are working on before the code freeze date, we'd really appreciate it.

_P.S: I am still working on the messaging for these bumps, so please let me know if you have any questions/concerns/feedback._

**Comment by pmusolino** [comment]
> yes, my device was set to March 29 in Rome/Italy for the screenshots. 23 Mar-29 Mar for "This week" tab looks correctly, if the device time is on March 29. since the first weekday starts on Monday in Rome/Italy calendar, it ends on Sunday (March 29).

Ohh I see, now it makes sense! 😄 

>  I just added a commit that passed the locale to the start/end of week date calculation and updated unit tests with locales that have different first weekday for another look!

Amazing! I take a look at the new code. 👍

File: RELEASE-NOTES.txt [hunk 1]
```diff
@@ -11,6 +11,7 @@
 - Fix pulling to refresh on the Processing tab sometimes will not show the up-to-date orders.
 - Edit Product > Price Settings: schedule sale is now available even when either the start or end date is not set, and the sale end date can be removed now.
 - Improved stats: fixed a crash when loading improved stats on a day when daily saving time changes.
+- Improved stats: fixed the incorrect time range on "This Week" tab when loading improved stats on a day when daily saving time changes.
```
Comment (pmusolino) [inline RELEASE-NOTES.txt:14]: We need to move this under the next release 4.2

Comment (jaclync) [reply RELEASE-NOTES.txt]: great catch! updated in https://github.com/woocommerce/woocommerce-ios/pull/2098/commits/7597543c41aa7818afdbc62f79ff5ced3f5906d5 🙂

**Pull request merged**